gateway/src/apicast/mapping_rule.lua

```diff
@@ -0,0 +1,94 @@
+local setmetatable = setmetatable
+local pairs = pairs
+local error = error
+local type = type
+local re_match = ngx.re.match
+local insert = table.insert
+
+local _M = {}
+
+local mt = { __index = _M }
+
+local function hash_to_array(hash)
+  local array = {}
+
+  for k,v in pairs(hash or {}) do
+    insert(array, { k, v })
+  end
+
+  return array
+end
+
+local function regexpify(path)
+  return path:gsub('?.*', ''):gsub("{.-}", '([\\w_.-]+)'):gsub("%.", "\\.")
+end
+
+local regex_variable = '\\{[-\\w_]+\\}'
+
+local function check_querystring_params(params, args)
+  local match = true
+
+  for i=1, #params do
+    local param = params[i][1]
+    local expected = params[i][2]
+    local m, err = re_match(expected, regex_variable, 'oj')
+    local value = args[param]
+
+    if m then
+      if not value then -- regex variable have to have some value
+        ngx.log(ngx.DEBUG, 'check query params ', param,
+          ' value missing ', expected)
+        match = false
+        break
+      end
+    else
+      if err then ngx.log(ngx.ERR, 'check match error ', err) end
+
+      -- if many values were passed use the last one
+      if type(value) == 'table' then
+        value = value[#value]
+      end
+
+      if value ~= expected then -- normal variables have to have exact value
+        ngx.log(ngx.DEBUG, 'check query params does not match ',
+          param, ' value ' , value, ' == ', expected)
+        match = false
+        break
+      end
+    end
+  end
+
+  return match
+end
+
+function _M.new(http_method, pattern, params, querystring_params, metric, delta)
```

Comment:
I'd keep this method private because there is high chance the initializer will change with that many params.

Reply:
Good point. Let's wait until we bring more functionality from `Service`.